proposals/0165-async-vote-execution.md

```diff
@@ -0,0 +1,192 @@
+validators can reach consensus faster.
+
+Even with async vote execution, forks can still happen because of
+various other situations, like network partitions or mis-configured validators.
+This work just reduces the chances of forks caused by variance in non-vote
+transaction executions.
+
+The non-vote transactions do need to be executed eventually. Even though it's
+hard to make sure everyone executes every block within 400ms, on average
+majority of the cluster should be able to keep up.
+
+## Alternatives Considered
+
+### Separating vote and non-vote transactions into different domains
+
+An earlier proposal of Async Execution proposes that we separate vote and
+non-vote transactions into different domains, so that we can execute them
+independently. The main concerns were:
+
+* We need to introduce one bit in AccountsDB for every account, this
+complicates the implementation
+
+* Topping off the vote fee payer accounts becomes difficult. We need to add a
+bounce account to move fees from user domain to vote domain, and the process
+may take one epoch
+
+## New Terminology
+
+* `Ephemeral Bankhash`: The hash calculated after executing only vote
+transactions in a block without checking fee payers. If there are no votes,
+use default hash.
+* `Final Bankhash`: The bankhash as we know it today. This is a hash calculated
+after executing all transactions in a block, checking fee payers for all.
+
+## Detailed Design
+
+### Allow leader to skip execution of transactions (Bankless Leader)
+
+There is already on-going effort to totally skip execution of all transactions
+when leader pack new blocks. See SIMD 82, SIMD 83, and related trackers:
+https://github.com/anza-xyz/agave/issues/2502
+
+Theoretically we could reap some benefit without Bankless Leader, the leader
+pack as normal, while other validators only replay votes first, then later
+execute other transactions and compare with the bankhash of the leader. But in
+such a setup we gain smaller speedup without much benefits, it is a possible
+route during rollouts though.
+
+### Calculate ephemeral hash executing votes only and vote on selected forks
+
+Two new fields will be added to `TowerSync` vote transaction:
+
+* `ephemeral_hash`: This is the hash obtained by hashing blockid and all
+vote transactions in the block together. Used primarily in consensus.
+* `ephemeral_slot`: This is the slot where the ephemeral_hash is calculated.
+
+This step is optimistic in the sense that validators do not check the fee
+payers when executing the vote transactions in a block. They assume vote
+transactions will not fail due to insufficient fees, apply the execution
+results to select the correct fork, then immediately vote on the bank with
+only the hash result of the vote transactions.
+
+This is safe because the list of validators and their corresponding stake
+has already been determined at the beginning of the Epoch. The stake we used
```

Comment:
The leader schedule for the current epoch was computed at the beginning of the previous epoch.  Not this epoch.

The fork choice stake weights should match the LS computation, otherwise there will be a stall waiting for async to finish at the start of the epoch.

Reply:
Done.